src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -33,107 +34,156 @@
-                field += csv.charAt(i++);
+            success = setType(pet, fields[2]);
+            success = setOwner(pet, fields[3]);
+            if(fields.length == 4){
+                i = updateDatabase(pet, fields[4]);
             }
-            i++;
+            pets.add(pet);
+        }
+        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    }
 
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    private int popNextField(int i, String field, String csv) {
+        while (i < csv.length() && csv.charAt(i) != ';') {
+            assert(csv.charAt(i) != '\n');
+            field += csv.charAt(i++);
+        }
+        assert(csv.charAt(i) != '\n');
+        assert(csv.charAt(i) == ';');
+        i++;
+        return i;
+    }
+
+    private int popNextFieldV2(int i, String field, String csv){
+        while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
+            field += csv.charAt(i++);
+        }
+        return i;
+    }
+
+    private int popNextFieldV3(int i, String field, String csv){
+        while (i < csv.length() && csv.charAt(i) != '\n') {
+            assert(csv.charAt(i) != ';');
+            field += csv.charAt(i++);
+        }
+        assert(csv.charAt(i) != ';');
+        assert(csv.charAt(i) == '\n');
+        return i;
+    }
+
+    private boolean areEqual(Pet onePet, Pet anotherPet) {
+        return onePet.getName().equals(anotherPet.getName()) &&
+            onePet.getType().getId().equals(anotherPet.getType().getId()) &&
+            onePet.getBirthDate().equals(anotherPet.getBirthDate());
+    }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
+    private int extractName(Pet pet, int i, String csv, String field) {
+        i = popNextField(i, field, csv);
+        i++;
+        return i;
+    }
+
+    private boolean setBirthDate(Pet pet, String value){
+        Date date;
+        try {
+            date = new SimpleDateFormat("yyyy-MM-dd").parse(value);
+        } catch (ParseException e) {
+            return false;
+        }
+        pet.setBirthDate(date);
+        return true;
+    }
+
+    private boolean setType(Pet pet, String value){
+        ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
+        for (int j = 0; j < ts.size(); j++) {
+            if (ts.get(j).getName().toLowerCase().equals(value)) {
+                pet.setType(ts.get(j));
+                return true;
             }
+        }
+        return false;
```

Comment:
If there is an invalid type, this call might actually fail. So we added a security check here.